pkg/manager/member/tikv_upgrader.go

```diff
@@ -14,26 +14,162 @@
+	tcName := tc.GetName()
 	if tc.Status.PD.Phase == v1alpha1.UpgradePhase {
 		_, podSpec, err := GetLastAppliedConfig(oldSet)
 		if err != nil {
 			return err
 		}
 		newSet.Spec.Template.Spec = *podSpec
-	} else {
-		tc.Status.TiKV.Phase = v1alpha1.UpgradePhase
+		return nil
+	}
+
+	if !tc.Status.TiKV.Synced {
+		return fmt.Errorf("Tidbcluster: [%s/%s]'s tikv status sync failed,can not to be upgraded", ns, tcName)
+	}
+
+	tc.Status.TiKV.Phase = v1alpha1.UpgradePhase
+	setUpgradePartition(newSet, *oldSet.Spec.UpdateStrategy.RollingUpdate.Partition)
+
+	if tc.Status.TiKV.StatefulSet.CurrentReplicas == 0 {
+		return controller.RequeueErrorf("tidbcluster: [%s/%s]'s tikv doesn't have old version pod to upgrade", ns, tcName)
+	}
+
+	if tc.Status.TiKV.StatefulSet.UpdatedReplicas+tc.Status.TiKV.StatefulSet.CurrentReplicas != tc.Status.TiKV.StatefulSet.Replicas {
+		return controller.RequeueErrorf("tidbcluster: [%s/%s]'s tikv pods are not all created", ns, tcName)
+	}
+	for i := tc.Status.TiKV.StatefulSet.Replicas; i > tc.Status.TiKV.StatefulSet.CurrentReplicas; i-- {
+		store := tku.getStoreByOrdinal(tc, i-1)
+		if store == nil || store.State != metapb.StoreState_name[int32(metapb.StoreState_Up)] {
+			return controller.RequeueErrorf("tidbcluster: [%s/%s]'s upgraded tikv pods are not all ready", ns, tcName)
+		}
+	}
+
+	upgradeOrdinal := tc.Status.TiKV.StatefulSet.CurrentReplicas - 1
+	upgradePodName := tikvPodName(tc, upgradeOrdinal)
+	upgradePod, err := tku.podLister.Pods(ns).Get(upgradePodName)
+	if err != nil {
+		return err
+	}
+
+	for _, store := range tc.Status.TiKV.TombstoneStores {
+		if store.PodName == upgradePodName {
+			setUpgradePartition(newSet, upgradeOrdinal)
+			return nil
+		}
+	}
+
+	for _, store := range tc.Status.TiKV.Stores {
+		if store.PodName == upgradePodName {
+			storeID, err := strconv.ParseUint(store.ID, 10, 64)
+			if err != nil {
+				return err
+			}
+			_, evicting := upgradePod.Annotations[EvictLeaderBeginTime]
+
+			if tku.readyToUpgrade(upgradePod, store) {
+				err = tku.pdControl.GetPDClient(tc).EndEvictLeader(storeID)
+				if err != nil {
+					return err
+				}
+				if evicting {
+					delete(upgradePod.Annotations, EvictLeaderBeginTime)
+					_, err = tku.podControl.UpdatePod(tc, upgradePod)
+					if err != nil {
+						return err
+					}
+				}
+				setUpgradePartition(newSet, upgradeOrdinal)
+				return nil
+			}
+
+			if !evicting {
+				err = tku.pdControl.GetPDClient(tc).BeginEvictLeader(storeID)
+				if err != nil {
+					return err
+				}
+				// updete upgradePod
+				if upgradePod.Annotations == nil {
+					upgradePod.Annotations = map[string]string{}
+				}
+				upgradePod.Annotations[EvictLeaderBeginTime] = time.Now().Format(time.RFC3339)
+				_, err = tku.podControl.UpdatePod(tc, upgradePod)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+func (tku *tikvUpgrader) readyToUpgrade(upgradePod *corev1.Pod, store v1alpha1.TiKVStore) bool {
```

Comment:
I think this method should have two return values: (bool, error)

Reply:
the error of the method is not needed to throw out.